odoo/models.py

```diff
@@ -6811,6 +6814,75 @@ def _invalidate_cache(self, fnames=None, ids=None):
                 spec.append((invf, None))
         self.env.cache.invalidate(spec)
 
+    @contextlib.contextmanager
+    def modifying_recordset(self, *, outputs: Iterable[str], inputs: Iterable[str] = (), discard_inputs: bool = False) -> Iterator[SimpleNamespace]:
+        """Helper for manipulating caches & computes around SQL queries.
+
+        Only affects the current recordset.
+
+        :param outputs: the list of fields modified by the SQL query, these
+            fields will be both :meth:`~.invalidate_recordset` and :meth:`~.modified`
+        :param inputs: the list of fields the SQL query needs / reads, these
+            fields will be :meth:`~.flush_recordset` unless `discard_inputs` is
+            set (in which case they are only invalidated)
+        :param discard_inputs: whether pending updates to the ``inputs`` should
+            be discarded / ignored, or applied before the query. Defaults to
+            ``False`` (any pending change to an input field will be flushed)
+        :return: a simple object whose `records` attribute can be updated to
+            control how many records are :meth:`~.modified`, useful if only a
+            subset of the recordset got updated
+        """
+        inputs = tuple(inputs)
+        outputs = tuple(outputs)
+        if discard_inputs:
+            self.invalidate_recordset(inputs + outputs, flush=False)
+        else:
+            if inputs:
+                self.invalidate_recordset(inputs, flush=True)
+            self.invalidate_recordset([f for f in outputs if f not in inputs], flush=False)
+
+        # alternative: raise a specific exception?
+        obj = SimpleNamespace(records=self)
+        yield obj
+        obj.records.modified(outputs)
+
+    @contextlib.contextmanager
+    def modifying_model(self, outputs: Iterable[str], inputs: Iterable[str] = (), discard_inputs: bool = False) -> Iterator[SimpleNamespace]:
```

Comment:
According to what I wrote above, if you modify relational fields, you must know in advance which records will be modified in order to determine all the records that need some recomputation.  Maybe there should be a safeguard here, to make sure we don't modify relational fields.  But such a safeguard is actually not modular: adding a relational computed field that depends on the modified fields will make the check fail.